tests/gtest/avifprogressivetest.cc

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "avif/avif.h"
+#include "aviftest_helpers.h"
+#include "gtest/gtest.h"
+
+namespace libavif {
+namespace {
+
+class ProgressiveTest : public testing::Test {
+ protected:
+  void SetUp() override {
+    if (avifCodecName(AVIF_CODEC_CHOICE_AOM, AVIF_CODEC_FLAG_CAN_ENCODE) ==
+        nullptr) {
+      GTEST_SKIP() << "ProgressiveTest requires the AOM encoder.";
+    }
+
+    ASSERT_NE(encoder_, nullptr);
+    encoder_->codecChoice = AVIF_CODEC_CHOICE_AOM;
+    // The fastest speed that uses AOM_USAGE_GOOD_QUALITY.
+    encoder_->speed = 6;
+
+    ASSERT_NE(decoder_, nullptr);
+    decoder_->allowProgressive = true;
+
+    ASSERT_NE(image_, nullptr);
+    testutil::FillImageGradient(image_.get());
+  }
+
+  void TestDecode(uint32_t expect_width, uint32_t expect_height) {
+    ASSERT_EQ(avifDecoderSetIOMemory(decoder_.get(), encoded_avif_.data,
+                                     encoded_avif_.size),
+              AVIF_RESULT_OK);
+    ASSERT_EQ(avifDecoderParse(decoder_.get()), AVIF_RESULT_OK);
+    ASSERT_EQ(decoder_->progressiveState, AVIF_PROGRESSIVE_STATE_ACTIVE);
+    ASSERT_EQ(static_cast<uint32_t>(decoder_->imageCount),
+              encoder_->extraLayerCount + 1);
+
+    for (uint32_t layer = 0; layer < encoder_->extraLayerCount + 1; ++layer) {
+      ASSERT_EQ(avifDecoderNextImage(decoder_.get()), AVIF_RESULT_OK);
+      // libavif scales frame automatically.
+      ASSERT_EQ(decoder_->image->width, expect_width);
+      ASSERT_EQ(decoder_->image->height, expect_height);
+      // TODO(wtc): Check avifDecoderNthImageMaxExtent().
+    }
+
+    // TODO(wtc): Check decoder_->image and image_ are similar, and better
+    // quality layer is more similar.
+  }
+
+  static constexpr uint32_t kImageSize = 256;
+
+  testutil::AvifEncoderPtr encoder_{avifEncoderCreate(), avifEncoderDestroy};
+  testutil::AvifDecoderPtr decoder_{avifDecoderCreate(), avifDecoderDestroy};
+
+  testutil::AvifImagePtr image_ =
+      testutil::CreateImage(kImageSize, kImageSize, 8, AVIF_PIXEL_FORMAT_YUV444,
+                            AVIF_PLANES_YUV, AVIF_RANGE_FULL);
+
+  testutil::AvifRwData encoded_avif_;
+};
+
+TEST_F(ProgressiveTest, QualityChange) {
+  encoder_->extraLayerCount = 1;
+  encoder_->minQuantizer = 50;
+  encoder_->maxQuantizer = 50;
+
+  ASSERT_EQ(avifEncoderAddImage(encoder_.get(), image_.get(), 1,
+                                AVIF_ADD_IMAGE_FLAG_NONE),
+            AVIF_RESULT_OK);
+
+  encoder_->minQuantizer = 0;
+  encoder_->maxQuantizer = 0;
+  ASSERT_EQ(avifEncoderAddImage(encoder_.get(), image_.get(), 1,
+                                AVIF_ADD_IMAGE_FLAG_NONE),
+            AVIF_RESULT_OK);
+
+  ASSERT_EQ(avifEncoderFinish(encoder_.get(), &encoded_avif_), AVIF_RESULT_OK);
+
+  TestDecode(kImageSize, kImageSize);
+}
+
+TEST_F(ProgressiveTest, DimensionChange) {
+  if (avifLibYUVVersion() == 0) {
+    GTEST_SKIP() << "libyuv not available, skip test.";
+  }
```

Comment:
@tongyuantongyu Yuan: I guess the reason we need to skip this test when libyuv is not available is because this test needs the `avifImageScale()` function, which is only implemented if libyuv is available. Correct?

Reply:
Yes.
I just noticed the assertion failure when using libaom head as well. Seems like a regression after v3.5.0 release.

Reply:
I have only tested with libaom v3.5.0 and v3.6.0-rc1. If it's a regression, I can do a git bisect.

Reply:
git bisect says:

```
fbebe9d771f485272d6dd9f3829d9389160d89e1 is the first bad commit
commit fbebe9d771f485272d6dd9f3829d9389160d89e1
Author: chiyotsai <chiyotsai@google.com>
Date:   Mon May 23 13:31:41 2022 -0700

    RESIZE_MODE: Fix incorrect strides being used for motion search
    
    Currently during motion search, the encoder always assumes that the ref
    frame's stride is the same as that of source frame. But this assumption
    breaks when resize mode/superres feature is turned on, leading to
    unexpected motion search result and invalid memory access.
    
    This commit actively checks for the ref stride with what's stored in
    MotionVectorSearchParams::search_site_config, and creates a new
    search_site_config if there is a mismatch.
    
    BUG=aomedia:3283
    
    Change-Id: Ia99a88326bf716027b5a652ec519fb13cfa0a345

 av1/av1.cmake                      |   1 +
 av1/encoder/block.h                |  10 ++++
 av1/encoder/encoder.c              |  13 -----
 av1/encoder/mcomp.c                |   2 +
 av1/encoder/mcomp.h                | 111 ++++++++++---------------------------
 av1/encoder/mcomp_structs.h        |  83 +++++++++++++++++++++++++++
 av1/encoder/motion_search_facade.c |  35 ++++++++----
 av1/encoder/motion_search_facade.h |  22 ++++++++
 av1/encoder/nonrd_pickmode.c       |   9 ++-
 av1/encoder/temporal_filter.c      |   5 +-
 10 files changed, 182 insertions(+), 109 deletions(-)
 create mode 100644 av1/encoder/mcomp_structs.h
```

But that is the commit that added the assertion that fails. I will ask Chi Yo (the author of that commit) about that assertion.